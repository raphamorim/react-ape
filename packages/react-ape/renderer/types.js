/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 *
 */

export type SpatialGeometry = {|
  x: number,
  y: number,
|};

export type ApeElement = {|
  render: (CanvasComponentContext, ?Layout) => mixed,
  clear?: () => mixed,
  parentLayout?: Layout
|};

export type Layout = {
  style?: {
    backgroundColor: string,
    borderColor: string,
  },
  spatialGeometry: SpatialGeometry
};

export type CanvasComponentContext = {
  // render queue for update operations
  renderQueue: Array<ApeElement>,
  type: 'canvas',
  ctx: CanvasRenderingContext2D,
};
