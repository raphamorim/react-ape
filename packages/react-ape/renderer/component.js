/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import Image from './elements/Image';
import Text from './elements/Text';
import View from './elements/View';
import {CustomComponents} from '../modules/Register';
import {insertNodeOnApeTree} from './apeTree/apeTree';
import {unsafeCreateUniqueId} from './utils';

const CHILDREN = 'children';
const STYLE = 'style';

const ReactApeComponent = {
  createElement(
    type,
    props,
    rootContainerElement,
    apeContextGlobal,
    internalInstanceHandle
  ) {
    // TODO: Run it once
    const customDict = {};
    Object.keys(CustomComponents).forEach(customKey => {
      customDict[customKey] = CustomComponents[customKey](
        props,
        apeContextGlobal
      );
    });

    const id = unsafeCreateUniqueId();

    const COMPONENTS = {
      ...customDict,
      Image: Image(props),
      Text: Text(props, id),
      View: new View(props),
    };

    if (!COMPONENTS[type]) {
      throw new Error(
        `React Ape could not identify "${type}" as ReactApeComponent. More details: http://raphamorim.io/react-ape`
      );
    }

    // TODO: rethink a better way of include it
    insertNodeOnApeTree(id, COMPONENTS[type]);
    COMPONENTS[type].id = id;

    return COMPONENTS[type];
  },

  createTextNode(text, rootContainerElement) {
    return text;
  },

  diffProperties(
    element,
    tag,
    lastRawProps,
    nextRawProps,
    rootContainerElement
  ) {
    let updatePayload = null;

    const lastProps = lastRawProps;
    const nextProps = nextRawProps;

    let propKey;
    let styleName;
    let styleUpdates = null;
    for (propKey in lastProps) {
      if (
        nextProps.hasOwnProperty(propKey) ||
        !lastProps.hasOwnProperty(propKey) ||
        lastProps[propKey] == null
      ) {
        continue;
      }
      if (propKey === STYLE) {
        const lastStyle = lastProps[propKey];
        for (styleName in lastStyle) {
          if (lastStyle.hasOwnProperty(styleName)) {
            if (!styleUpdates) {
              styleUpdates = {};
            }
            styleUpdates[styleName] = '';
          }
        }
      } else {
        // For all other deleted properties we add it to the queue. We use
        // the whitelist in the commit phase instead.
        (updatePayload = updatePayload || []).push(propKey, null);
      }
    }
    for (propKey in nextProps) {
      const nextProp = nextProps[propKey];
      const lastProp = lastProps != null ? lastProps[propKey] : undefined;
      if (
        !nextProps.hasOwnProperty(propKey) ||
        nextProp === lastProp ||
        (nextProp == null && lastProp == null)
      ) {
        continue;
      }
      if (propKey === STYLE) {
        if (lastProp) {
          // Unset styles on `lastProp` but not on `nextProp`.
          for (styleName in lastProp) {
            if (
              lastProp.hasOwnProperty(styleName) &&
              (!nextProp || !nextProp.hasOwnProperty(styleName))
            ) {
              if (!styleUpdates) {
                styleUpdates = {};
              }
              styleUpdates[styleName] = '';
            }
          }
          // Update styles that changed since `lastProp`.
          for (styleName in nextProp) {
            if (
              nextProp.hasOwnProperty(styleName) &&
              lastProp[styleName] !== nextProp[styleName]
            ) {
              if (!styleUpdates) {
                styleUpdates = {};
              }
              styleUpdates[styleName] = nextProp[styleName];
            }
          }
        } else {
          // Relies on `updateStylesByID` not mutating `styleUpdates`.
          if (!styleUpdates) {
            if (!updatePayload) {
              updatePayload = [];
            }
            updatePayload.push(propKey, styleUpdates);
          }
          styleUpdates = nextProp;
        }
      } else if (propKey === CHILDREN) {
        if (
          lastProp !== nextProp &&
          (typeof nextProp === 'string' || typeof nextProp === 'number')
        ) {
          (updatePayload = updatePayload || []).push(propKey, nextProp);
        }
      } else {
        // For any other property we always add it to the queue and then we
        // filter it out using the whitelist during the commit.
        (updatePayload = updatePayload || []).push(propKey, nextProp);
      }
    }
    if (styleUpdates) {
      (updatePayload = updatePayload || []).push(STYLE, styleUpdates);
    }

    return updatePayload;
  },
};

export default ReactApeComponent;
