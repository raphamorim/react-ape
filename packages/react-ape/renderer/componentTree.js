/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {unsafeCreateUniqueId} from './utils';

const randomKey = unsafeCreateUniqueId();
const internalInstanceKey = '__reactInternalInstance$' + randomKey;
const internalEventHandlersKey = '__reactEventHandlers$' + randomKey;

export function precacheFiberNode(hostInst, node) {
  node[internalInstanceKey] = hostInst;
}

export function getFiberCurrentPropsFromNode(node) {
  return node[internalEventHandlersKey] || null;
}

export function updateFiberProps(node, props) {
  node[internalEventHandlersKey] = props;
}
