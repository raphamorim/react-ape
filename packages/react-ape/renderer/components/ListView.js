/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 *
 */

import * as React from 'react';

/**
  Usage:

  <ListView
    dataSource={[]}
    renderRow={() => {}}
    style={StyleObject}
  />
*/

type Props = {|
  style: {[string]: string | number},
  dataSource: Array<mixed>,
  renderRow: mixed => React.Node,
|};

class ListView extends React.Component<Props> {
  static defaultProps: Props = {
    renderRow: () => null,
    dataSource: [],
    style: {},
  };

  render(): React.Element<string> {
    const {style, renderRow, dataSource} = this.props;
    return React.createElement(
      'View',
      {style},
      dataSource.length ? dataSource.map(renderRow) : null
    );
  }
}

export default ListView;
