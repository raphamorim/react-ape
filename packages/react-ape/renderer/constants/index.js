/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

// Defaults for Render
export const ViewDefaults = {
  size: 100, // 200x200
  lineHeight: 24,
};

// ReactApe Internal Constants
export const _SectionBlockSize: number = 80; // 80x80

// DevTools configuration
export const DevToolsConfig = {
  // $FlowFixMe[signature-verification-failure]
  bundleType: process.env.NODE_ENV === 'production' ? 0 : 1,
  version: '0.1.0',
  rendererPackageName: 'ReactApe',
};
