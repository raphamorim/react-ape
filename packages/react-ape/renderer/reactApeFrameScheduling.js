// @flow
// https://github.com/facebook/react/blob/master/packages/react-native-renderer/src/ReactNativeFrameScheduling.js
// TODO: make this hardware relevant/aware

const hasNativePerformanceNow =
  typeof performance === 'object' && typeof performance.now === 'function';

export const now = hasNativePerformanceNow
  ? () => performance.now()
  : () => Date.now();

let scheduledCallback: (() => mixed) | null = null;
let frameDeadline: number = 0;

function setTimeoutCallback() {
  // TODO (bvaughn) Hard-coded 5ms unblocks initial async testing.
  // React API probably changing to boolean rather than time remaining.
  // Longer-term plan is to rewrite this using shared memory,
  // And just return the value of the bit as the boolean.
  frameDeadline = now() + 5;

  const callback = scheduledCallback;
  scheduledCallback = null;
  if (callback !== null) {
    callback();
  }
}

// RN has a poor polyfill for requestIdleCallback so we aren't using it.
// This implementation is only intended for short-term use anyway.
// We also don't implement cancel functionality b'c Fiber doesn't currently need it.
export function scheduleDeferredCallback(
  callback: () => mixed,
  options?: {timeout: number},
): number {
  // We assume only one callback is scheduled at a time b'c that's how Fiber works.
  scheduledCallback = callback;
  const timeoutId = setTimeout(setTimeoutCallback, 1);
  return (timeoutId: any); // Timeouts are always numbers on RN
}

export function cancelDeferredCallback(callbackID: number) {
  scheduledCallback = null;
  clearTimeout((callbackID: any)); // Timeouts are always numbers on RN
}

export function shouldYield() {
  return frameDeadline <= now();
}
