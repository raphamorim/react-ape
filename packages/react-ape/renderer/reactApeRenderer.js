/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {CanvasComponentContext} from './types';

import reconciler from 'react-reconciler';
import reactApeComponent from './reactApeComponent';
import {scaleDPI, clearCanvas} from './core/canvas';
import {renderElement, renderQueue} from './core/render';
import {precacheFiberNode, updateFiberProps} from './reactApeComponentTree';
import devToolsConfig from './config/devtools';
import {
  now as FrameSchedulingNow,
  cancelDeferredCallback as FrameSchedulingCancelDeferredCallback,
  scheduleDeferredCallback as FrameSchedulingScheduleDeferredCallback,
  shouldYield as FrameSchedulingShouldYield,
} from './reactApeFrameScheduling';

// TODO: Use Context.
let apeContextGlobal = null;
let surfaceHeight = 0;

const ReactApeFiber = reconciler({
  appendInitialChild(parentInstance, child) {
    if (parentInstance.appendChild && child.type !== 'View') {
      parentInstance.appendChild(child);

      // TODO: Change it later
      child.getParentLayout = parentInstance.getLayoutDefinitions;
    }
  },

  createInstance(
    type,
    props,
    rootContainerInstance,
    hostContext,
    internalInstanceHandle
  ) {
    if (!apeContextGlobal && rootContainerInstance.getContext) {
      const rootContainerInstanceContext = rootContainerInstance.getContext('2d');

      // TODO: Change it.
      scaleDPI(rootContainerInstance, rootContainerInstanceContext);

      apeContextGlobal = {
        type: 'canvas',
        getSurfaceHeight: () => surfaceHeight,
        setSurfaceHeight: height => {
          surfaceHeight = height;
        },
        renderQueue: [],
        ctx: rootContainerInstanceContext,
      };
    }

    const apeElement = reactApeComponent.createElement(
      type,
      props,
      rootContainerInstance,
      apeContextGlobal,
      internalInstanceHandle
    );

    precacheFiberNode(internalInstanceHandle, apeElement);
    updateFiberProps(apeElement, props);
    return apeElement;
  },

  createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
    // let textNode = reactApeComponent.createTextNode(text, rootContainerInstance);
    // precacheFiberNode(internalInstanceHandle, textNode);
    return text;
  },

  finalizeInitialChildren(parentInstance, type, props) {
    if (type === 'View') {
      parentInstance.render(apeContextGlobal);
    }
    return false;
  },

  getPublicInstance(inst) {
    return inst;
  },

  prepareForCommit(rootContainerInstance) {
  },

  prepareUpdate(element, type, oldProps, newProps, rootContainerInstance) {
    if (newProps) {
      const diff = reactApeComponent.diffProperties(
        element,
        type,
        oldProps,
        newProps,
        rootContainerInstance
      );

      if (diff) {
        const parentLayout = element.parentLayout || element.getParentLayout();
        element.clear(oldProps, parentLayout, apeContextGlobal);
        const {style = {}} = oldProps;

        const apeElement = reactApeComponent.createElement(
          type,
          newProps,
          rootContainerInstance,
          apeContextGlobal
        );

        if (diff.indexOf('children') === -1) {
          renderElement(apeContextGlobal, apeElement, parentLayout);
          return;
        }

        apeElement.parentLayout = parentLayout;
        apeContextGlobal.renderQueue.push(apeElement);
      }
    }
  },

  resetAfterCommit(rootContainerInstance) {
    // resetAfterCommit happens only for children changes
    renderQueue(apeContextGlobal, () => {
      apeContextGlobal.setSurfaceHeight(0);
      apeContextGlobal.renderQueue = [];
    });
  },

  resetTextContent(element) {
    // noop
  },

  getRootHostContext(rootInstance) {
    return {};
  },

  getChildHostContext() {
    return {};
  },

  scheduleTimeout: setTimeout,
  cancelTimeout: clearTimeout,
  scheduleDeferredCallback: FrameSchedulingScheduleDeferredCallback,
  cancelDeferredCallback: FrameSchedulingCancelDeferredCallback,
  schedulePassiveEffects: FrameSchedulingScheduleDeferredCallback,
  cancelPassiveEffects: FrameSchedulingCancelDeferredCallback,
  noTimeout: -1,
  useSyncScheduling: true,
  now: FrameSchedulingNow,

  isPrimaryRenderer: true,
  supportsMutation: true,

  shouldDeprioritizeSubtree(type, props) {
    return false;
  },

  appendChildToContainer(parentInstance, child) {
    if (child.render) {
      child.render(apeContextGlobal);
    }
  },

  appendChildToContainer(parentInstance, child) {
    if (child.render) {
      child.render(apeContextGlobal);
    }
  },

  appendChild(parentInstance, child) {
    // console.log(parentInstance, child);
    // if (parentInstance.appendChild) {
    //   parentInstance.appendChild(child);
    // }
  },

  removeChild(parentInstance, child) {
    // parentInstance.removeChild(child);
  },

  removeChildFromContainer(parentInstance, child) {
    // parentInstance.removeChild(child);
  },

  insertInContainerBefore(parentInstance, child, beforeChild) {
    // noop
  },

  commitUpdate(instance, updatePayload, type, oldProps, newProps) {
    // noop
  },

  commitMount(instance, updatePayload, type, oldProps, newProps) {
    // noop
  },

  commitTextUpdate(textInstance, oldText, newText) {
    // textInstance.children = newText;
  },

  shouldSetTextContent(props) {
    return (
      typeof props.children === 'string' || typeof props.children === 'number'
    );
  },
});

ReactApeFiber.injectIntoDevTools({
  ...devToolsConfig,
  findHostInstanceByFiber: ReactApeFiber.findHostInstance,
});

const defaultContainer = {};
// Using WeakMap avoids memory leak in case the container is garbage colected.
const roots = typeof WeakMap === 'function' ? new WeakMap() : new Map();

const ReactApeRenderer = {
  render(reactApeElement, canvasDOMElement, callback) {
    const containerKey =
      canvasDOMElement == null ? defaultContainer : canvasDOMElement;
    let root = roots.get(containerKey);
    if (!root) {
      root = ReactApeFiber.createContainer(containerKey);
      roots.set(canvasDOMElement, root);
      apeContextGlobal = null;
    }

    ReactApeFiber.updateContainer(reactApeElement, root, null, callback);

    return ReactApeFiber.getPublicRootInstance(root);
  },
};

export default ReactApeRenderer;
