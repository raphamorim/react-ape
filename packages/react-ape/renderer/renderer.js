/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {CanvasComponentContext} from './types';

import reconciler from 'react-reconciler';
import reactApeComponent from './component';
import {DevToolsConfig} from './constants';
import {scaleDPI} from './core/canvas';
import {renderElement, renderQueue} from './core/render';
import {precacheFiberNode, updateFiberProps} from './componentTree';
import {associateNodeOnApeTree} from './apeTree/apeTree';
import {
  now as FrameSchedulingNow,
  cancelDeferredCallback as FrameSchedulingCancelDeferredCallback,
  scheduleDeferredCallback as FrameSchedulingScheduleDeferredCallback,
  shouldYield as FrameSchedulingShouldYield,
} from './frameScheduling';

let apeContextGlobal = null;
let surfaceHeight = 0;

const ReactApeFiber = reconciler({
  appendInitialChild(parentInstance, child) {
    // if (parentInstance.appendChild && child.type !== 'View') {
    //   // START-TODO: delete it
    //   let layout = {};
    //   if (child.instructions && child.instructions.relative) {
    //     layout = {
    //       ...layout,
    //       ...parentInstance.getAndUpdateCurrentLayout(),
    //     };
    //   }
    //   parentInstance.appendChild({...child, layout});
    //   child.getParentLayout = parentInstance.getLayoutDefinitions;
    //   // END-TODO
    // }

    // TODO: it ended up adding 3 views: c(a,b,c) a(b) b(c)
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    }

    associateNodeOnApeTree(child.id, parentInstance.id);
  },

  createInstance(
    type,
    props,
    rootContainerInstance,
    hostContext,
    internalInstanceHandle
  ) {
    if (!apeContextGlobal && rootContainerInstance.getContext) {
      const rootContainerInstanceContext = rootContainerInstance.getContext(
        '2d'
      );

      scaleDPI(rootContainerInstance, rootContainerInstanceContext);
      apeContextGlobal = {
        getSurfaceHeight: () => surfaceHeight,
        setSurfaceHeight: height => {
          surfaceHeight = height;
        },
        ctx: rootContainerInstanceContext,
        renderQueue: [],
      };
    }

    const apeElement = reactApeComponent.createElement(
      type,
      props,
      rootContainerInstance,
      apeContextGlobal,
      internalInstanceHandle
    );

    precacheFiberNode(internalInstanceHandle, apeElement);
    updateFiberProps(apeElement, props);
    return apeElement;
  },

  createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
    // let textNode = reactApeComponent.createTextNode(text, rootContainerInstance);
    // precacheFiberNode(internalInstanceHandle, textNode);
    return text;
  },

  finalizeInitialChildren(parentInstance, type, props) {
    // Ele renderiza text, view, view (de baixo pra cima)
    return false;
  },

  getPublicInstance(inst) {
    return inst;
  },

  prepareForCommit(rootContainerInstance) {},

  prepareUpdate(element, type, oldProps, newProps, rootContainerInstance) {
    if (newProps) {
      const diff = reactApeComponent.diffProperties(
        element,
        type,
        oldProps,
        newProps,
        rootContainerInstance
      );

      if (diff) {
        const parentLayout = element.parentLayout || element.getParentLayout();
        if (type === 'Text') {
          parentLayout.resetLayout(); // View needs to be reset on text updates
        }

        element.clear(oldProps, parentLayout, apeContextGlobal);

        const apeElement = reactApeComponent.createElement(
          type,
          newProps,
          rootContainerInstance,
          apeContextGlobal
        );

        apeElement.parentLayout = parentLayout;

        // If isn't a children update, should render with new props
        if (diff.length && diff.indexOf('children') === -1) {
          renderElement(apeContextGlobal, apeElement, parentLayout);
          return null;
        }

        apeContextGlobal.renderQueue.push(apeElement);
        return null;
      }

      if (type === 'Text' && newProps.children && newProps.children.join) {
        const parentLayout = element.parentLayout || element.getParentLayout();
        element.clear(oldProps, parentLayout, apeContextGlobal);

        const apeElement = reactApeComponent.createElement(
          type,
          {...newProps, children: newProps.children.join('')},
          rootContainerInstance,
          apeContextGlobal
        );

        renderElement(apeContextGlobal, apeElement, parentLayout);
      }
    }
  },

  clearContainer() {},

  resetAfterCommit(rootContainerInstance) {
    // resetAfterCommit happens only for children changes
    renderQueue(apeContextGlobal, () => {
      apeContextGlobal.setSurfaceHeight(0);
    });
  },

  resetTextContent(element) {
    // noop
  },

  getRootHostContext(rootInstance) {
    return {
      // TODO: ADICIONAR COISAS AKI
    };
  },

  getChildHostContext() {
    return {};
  },

  scheduleTimeout: setTimeout,
  cancelTimeout: clearTimeout,
  scheduleDeferredCallback: FrameSchedulingScheduleDeferredCallback,
  cancelDeferredCallback: FrameSchedulingCancelDeferredCallback,
  schedulePassiveEffects: FrameSchedulingScheduleDeferredCallback,
  cancelPassiveEffects: FrameSchedulingCancelDeferredCallback,
  noTimeout: -1,
  useSyncScheduling: false,
  now: FrameSchedulingNow,
  isPrimaryRenderer: true,
  supportsMutation: true,

  shouldDeprioritizeSubtree(type, props) {
    return false;
  },

  appendChildToContainer(container, child) {
    // It goes in the container and append each child
    // apeContextGlobal.setSurfaceHeight(0);
    if (child.render) {
      child.render(apeContextGlobal);
    }
  },

  appendChild(parentInstance, child) {
    // console.log(3);
    // if (parentInstance.appendChild) {
      // parentInstance.appendChild(child);
    // }
    // console.log(2, parentInstance, child)
  },

  removeChild(parentInstance, child) {
    // parentInstance.removeChild(child);
    if (child.type && child.type === 'View') {
      child.clear(apeContextGlobal, parentInstance.getLayoutDefinitions());
    }
  },

  removeChildFromContainer(parentInstance, child) {
    // parentInstance.removeChild(child);
  },

  insertInContainerBefore(parentInstance, child, beforeChild) {
    // noop
  },

  commitUpdate(instance, updatePayload, type, oldProps, newProps) {
    // noop
  },

  commitMount(instance, updatePayload, type, oldProps, newProps) {
    // noop
  },

  commitTextUpdate(textInstance, oldText, newText) {
    // textInstance.children = newText;
    // console.log(111, textInstance, oldText, newText);
  },

  shouldSetTextContent(props) {
    return (
      typeof props.children === 'string' || typeof props.children === 'number'
    );
  },
});

ReactApeFiber.injectIntoDevTools({
  ...DevToolsConfig,
  findHostInstanceByFiber: ReactApeFiber.findHostInstance,
});

const defaultContainer = {};
// Using WeakMap avoids memory leak in case the container is garbage colected.
const roots = typeof WeakMap === 'function' ? new WeakMap() : new Map();

const ReactApeRenderer = {
  render(reactApeElement, canvasDOMElement, callback) {
    const containerKey =
      canvasDOMElement == null ? defaultContainer : canvasDOMElement;
    let root = roots.get(containerKey);
    if (!root) {
      root = ReactApeFiber.createContainer(containerKey);
      roots.set(canvasDOMElement, root);
      apeContextGlobal = null;
    }

    ReactApeFiber.updateContainer(reactApeElement, root, null, callback);

    return ReactApeFiber.getPublicRootInstance(root);
  },
};

export default ReactApeRenderer;
