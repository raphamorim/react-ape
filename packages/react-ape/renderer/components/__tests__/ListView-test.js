import React from 'react';
import renderer from 'react-test-renderer';

import ListView from '../ListView';
import {View, Text} from '../../../entry';

describe('ListView', () => {
  it("should render empty view when doesn't exist dataSource", () => {
    const ListViewTree = renderer.create(<ListView />).toJSON();
    expect(ListViewTree).toMatchSnapshot(`<View/>`);
  });

  it('should render properly with dataSource and renderRow', () => {
    const dataSource = [
      {dog: 'Pug', age: 5},
      {dog: 'Golden Retriever', age: 8},
    ];
    const renderRow = (data, idx) =>
      <View key={idx}>
        <Text>
          {data.dog}, which age is {data.age}
        </Text>
      </View>;

    const ListViewTree = renderer
      .create(<ListView renderRow={renderRow} dataSource={dataSource} />)
      .toJSON();

    expect(ListViewTree).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const dataSource = [{name: 'Jack'}, {name: 'Russel'}];
    const renderRow = (data, idx) =>
      <Text key={idx} id={'render-row-' + idx}>
        {data.name}
      </Text>;

    const ListViewTree = renderer
      .create(<ListView renderRow={renderRow} dataSource={dataSource} />)
      .toJSON();

    expect(ListViewTree).toMatchSnapshot();
  });
});
