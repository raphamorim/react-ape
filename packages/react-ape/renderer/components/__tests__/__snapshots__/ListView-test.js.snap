// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView renders correctly 1`] = `
<View>
  <Text
    id="render-row-0"
  >
    Jack
  </Text>
  <Text
    id="render-row-1"
  >
    Russel
  </Text>
</View>
`;

exports[`ListView should render empty view when doesn't exist dataSource: <View/> 1`] = `<View />`;

exports[`ListView should render properly with dataSource and renderRow 1`] = `
<View>
  <View>
    <Text>
      Pug
      , which age is 
      5
    </Text>
  </View>
  <View>
    <Text>
      Golden Retriever
      , which age is 
      8
    </Text>
  </View>
</View>
`;
